### PR TITLE
Don't run push mirrors for archived repos

### PR DESCRIPTION
**Pull request by JakobDev**
Fixes https://codeberg.org/forgejo/forgejo/issues/612

At the moment push mirrors are still run if a repo is archived. This PR fixes this.

### CODE REVIEW

**Comment by KN4CK3R** [comment]
Don't know if that is really not the preferred behaviour. If you add a new push mirror for an archived repo I would expect that the mirrored-to repo gets updated.

**Comment by wxiaoguang** [comment]
Considering an archived repo is read-only, I don't see it should support adding more push mirrors or do the push.

I am neutral for this change, but the nit is that I think the "check" could be done in the "handler" (after m.GetRepository()), then the SQL could still be kept simple, and the code could be easier to read and it's easier to add test code for it (maybe in the future).

**Comment by lunny** [comment]
I aslo think it's unnecessary to push to upstream when a repository has become archived one. Since there will be no new commits pushed to this archived repository and it will always keep the last commits not changed. Why it should keep to push to remote? It's wasting cpu and memory.

I think the right way to fix this is to disable the buttons on the UI and API to disallow change the push mirrors information. These push mirrors should not be removed because maybe it will be converted back to common mirror repositories.

**Comment by JakobDev** [comment]
> I think the right way to fix this is to disable the buttons on the UI and API to disallow change the push mirrors information. 

Done

grafik

> ~~but the nit is that I think the "check" could be done in the "handler" (after m.GetRepository()), then the SQL could still be kept simple, and the code could be easier to read and it's easier to add test code for it (maybe in the future).~~

I think the SQL is simple enough. Joins are just basic SQL that everyone who knows SQL should understand. We should not load more from the Database than needed.

> Another question is that what's the behavior of the "pull mirror"? Should it pull for archived repo?

You can't archive a pull mirror.

grafik

**Comment by lunny** [comment]
> > I think the right way to fix this is to disable the buttons on the UI and API to disallow change the push mirrors information.
> 
> Done
> 
> grafik
> 
> > ~but the nit is that I think the "check" could be done in the "handler" (after m.GetRepository()), then the SQL could still be kept simple, and the code could be easier to read and it's easier to add test code for it (maybe in the future).~
> 
> I think the SQL is simple enough. Joins are just basic SQL that everyone who knows SQL should understand. We should not load more from the Database than needed.
> 
> > Another question is that what's the behavior of the "pull mirror"? Should it pull for archived repo?
> 
> You can't archive a pull mirror.
> 
> grafik

Could we just disable the mirrors' buttons but keep the information displayed?

**Review by wxiaoguang — DISMISSED** [review]
Overall LGTM.

I would suggest not indenting the `{{else}}` block, to avoid unnecessary conflicts.

**Comment by 6543** [comment]
Please resolve conflict :)

File: models/repo/pushmirror.go [hunk 1]
```diff
@@ -137,8 +137,11 @@ func GetPushMirrorsSyncedOnCommit(ctx context.Context, repoID int64) ([]*PushMir
 // PushMirrorsIterate iterates all push-mirror repositories.
 func PushMirrorsIterate(ctx context.Context, limit int, f func(idx int, bean any) error) error {
 	sess := db.GetEngine(ctx).
-		Where("last_update + (`interval` / ?) <= ?", time.Second, time.Now().Unix()).
-		And("`interval` != 0").
+		Table("push_mirror").
+		Join("INNER", "`repository`", "`repository`.id = `push_mirror`.repo_id").
+		Where("`push_mirror`.last_update + (`push_mirror`.`interval` / ?) <= ?", time.Second, time.Now().Unix()).
+		And("`push_mirror`.`interval` != 0").
+		And("`repository`.is_archived = false").
```
Comment (delvh) [inline models/repo/pushmirror.go]: I think this needs to be
```suggestion
		And("`repository`.is_archived = ?", false).
```
as some DBs store booleans as ints.

**Review by delvh — APPROVED** [review]

**Comment by denyskon** [comment]
@JakobDev Please resolve conflicts :)

**Review by denyskon — APPROVED** [review]

**Pull request merged**